File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        visited = set()
        result = False
        def pre_order(node , target):
            nonlocal result
            target -= node.val
            if not node.left and not node.right and target == 0:
                result = True
                return 
            if node.left :
                pre_order(node.left , target)
            if node.right:
                pre_order(node.right , target)
            return
        if not root : return False 
        pre_order(root,targetSum)
        return result



            
                
       

            
        

File: 0112-path-sum/0112-path-sum.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


bool cal(struct TreeNode * node , int target){
    if(!node)
        return false;
    if(node -> left == NULL && node -> right == NULL && target == node -> val)
        return true;
    bool left = cal(node -> left , target - node -> val );
    bool right = cal(node -> right , target - node -> val);
    return left || right ;


    
    
    
}
bool hasPathSum(struct TreeNode* root, int targetSum) {
    
    return cal(root,targetSum);
    

    
}